Geometry/geometryStd_class.f90

```diff
@@ -202,24 +237,42 @@ subroutine move_noCache(self, coords, maxDist, event)
     type(coordList), intent(inout) :: coords
     real(defReal), intent(inout)   :: maxDist
     integer(shortInt), intent(out) :: event
-    integer(shortInt)              :: surfIdx, level
-    real(defReal)                  :: dist
+    integer(shortInt)              :: surfIdx, level, level0
+    real(defReal)                  :: dist, fieldDist
     class(surface), pointer        :: surf
     class(universe), pointer       :: uni
-    character(100), parameter :: Here = 'move (geometryStd_class.f90)'
+    character(100), parameter :: Here = 'move_noCache (geometryStd_class.f90)'
 
     if (.not.coords % isPlaced()) then
       call fatalError(Here, 'Coordinate list is not placed in the geometry')
     end if
 
+    level0 = coords % nesting
+    
     ! Find distance to the next surface
     call self % closestDist(dist, surfIdx, level, coords)
+    
+    ! Check fields
+    fieldDist = INF
+    if (allocated(self % temperatureField)) then
+      fieldDist = min(fieldDist, self % temperatureField % distance(coords))
+    end if
 
-    if (maxDist < dist) then ! Moves within cell
+    if (allocated(self % densityField)) then
+      fieldDist = min(fieldDist, self % densityField % distance(coords))
+    end if
+
+    if (maxDist < dist .and. maxDist < fieldDist) then ! Moves within cell
       call coords % moveLocal(maxDist, coords % nesting)
       event = COLL_EV
       maxDist = maxDist ! Left for explicitness. Compiler will not stand it anyway
-
+    
+    ! This check is really awful - can we do something better?
+    else if (fieldDist < dist .and. abs(fieldDist - dist) > 10*NUDGE) then ! Stays within the same cell, but crosses field boundary
```

Comment:
Can't you just do 
else if (maxDist < dist .and. maxDist >= fieldDist)?
in this case you're not crossing any geometrical boundary but a field boundary.